protocols/v2/roles-logic-sv2/src/channel_logic/channel_factory.rs

```diff
@@ -0,0 +1,1075 @@
+        }
+    }
+
+    pub fn get_version(&self) -> u32 {
+        match self {
+            Share::Extended(s) => s.version,
+            Share::Standard(s) => s.0.version,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ChannelFactory {
+    ids: Arc<Mutex<GroupId>>,
+    standard_channels_for_non_hom_downstreams: HashMap<u64, StandardChannel>,
+    standard_channels_for_hom_downstreams: HashMap<u32, StandardChannel>,
+    extended_channels: HashMap<u32, OpenExtendedMiningChannelSuccess<'static>>,
+    extranonces: ExtendedExtranonce,
+    share_per_min: f32,
+    // (NewExtendedMiningJob,group ids that already received the future job)
+    future_jobs: Vec<(NewExtendedMiningJob<'static>, Vec<u32>)>,
+    // (SetNewPrevHash,group ids that already received the set prev_hash)
+    last_prev_hash: Option<(SetNewPrevHash<'static>, Vec<u32>)>,
+    last_prev_hash_: Option<bitcoin::hash_types::BlockHash>,
+    // (NewExtendedMiningJob,group ids that already received the job)
+    last_valid_job: Option<(NewExtendedMiningJob<'static>, Vec<u32>)>,
+    job_creator: JobsCreators,
+    kind: ExtendedChannelKind,
+    job_ids: Id,
+}
+
+pub type PoolChannelFactory = ChannelFactory;
+
+impl ChannelFactory {
+    pub fn new(
+        ids: Arc<Mutex<GroupId>>,
+        extranonces: ExtendedExtranonce,
+        job_creator: JobsCreators,
+        share_per_min: f32,
+        kind: ExtendedChannelKind,
+    ) -> Self {
+        Self {
+            ids,
+            standard_channels_for_non_hom_downstreams: HashMap::new(),
+            standard_channels_for_hom_downstreams: HashMap::new(),
+            extended_channels: HashMap::new(),
+            extranonces,
+            share_per_min,
+            future_jobs: Vec::new(),
+            last_prev_hash: None,
+            last_prev_hash_: None,
+            last_valid_job: None,
+            job_creator,
+            kind,
+            job_ids: Id::new(),
+        }
+    }
+
+    pub fn add_standard_channel(
+        &mut self,
+        request_id: u32,
+        downstream_hash_rate: f32,
+        is_header_only: bool,
+        id: u32,
+    ) -> Result<Vec<Mining>, Error> {
+        match is_header_only {
+            true => {
+                self.new_standard_channel_for_hom_downstream(request_id, downstream_hash_rate, id)
+            }
+            false => self.new_standard_channel_for_non_hom_downstream(
+                request_id,
+                downstream_hash_rate,
+                id,
+            ),
+        }
+    }
+
+    pub fn new_extended_channel(
+        &mut self,
+        request_id: u32,
+        hash_rate: f32,
+        min_extranonce_size: u16,
+    ) -> Option<Vec<Mining>> {
+        let extended_channels_group = 0;
+        let max_extranonce_size = self.extranonces.get_range2_len() as u16;
+        if min_extranonce_size <= max_extranonce_size {
+            // SECURITY is very unlikely to finish the ids btw this unwrap could be used by an attaccher that
```

Comment:
Yeah - I could see this potentially causing a problem - it looks like the id would just wrap and start using/reusing/stepping on other ids. I made an issue to look into this at some point - https://github.com/stratum-mining/stratum/issues/379

Note: it'd take 136 years if we were using 1 id/second to reach max :)

Reply:
ok for now we have an issue for that and we leave it like as it is in this PR